pkg/pool-cow/contracts/CowRouter.sol

```diff
@@ -0,0 +1,349 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.24;
+
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeCast } from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+
+import { ICowRouter } from "@balancer-labs/v3-interfaces/contracts/pool-cow/ICowRouter.sol";
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+import {
+    AddLiquidityKind,
+    AddLiquidityParams,
+    SwapKind,
+    VaultSwapParams
+} from "@balancer-labs/v3-interfaces/contracts/vault/VaultTypes.sol";
+
+import { FixedPoint } from "@balancer-labs/v3-solidity-utils/contracts/math/FixedPoint.sol";
+
+import { VaultGuard } from "@balancer-labs/v3-vault/contracts/VaultGuard.sol";
+import { SingletonAuthentication } from "@balancer-labs/v3-vault/contracts/SingletonAuthentication.sol";
+
+contract CowRouter is SingletonAuthentication, VaultGuard, ICowRouter {
+    using FixedPoint for uint256;
+    using SafeERC20 for IERC20;
+    using SafeCast for *;
+
+    // Protocol fee percentage capped at 10%.
+    uint256 internal constant _MAX_PROTOCOL_FEE_PERCENTAGE = 10e16;
+
+    uint256 internal _protocolFeePercentage;
+    mapping(IERC20 => uint256) internal _protocolFees;
+
+    constructor(IVault vault, uint256 protocolFeePercentage) VaultGuard(vault) SingletonAuthentication(vault) {
+        _protocolFeePercentage = protocolFeePercentage;
```

Comment:
Since it has validation and an event, use the `_setProtocolFeePercentage` pattern and call it from here and in the external setter.